scylla/src/transport/locator/replication_info.rs

```diff
@@ -0,0 +1,183 @@
+            global_ring,
+            unique_nodes_in_global_ring,
+            datacenters,
+            empty_dc_data,
+        }
+    }
+
+    pub fn simple_strategy_replicas(
+        &self,
+        token: impl Borrow<Token>,
+        mut replication_factor: usize,
+    ) -> impl Iterator<Item = &Arc<Node>> {
+        if replication_factor > self.unique_nodes_in_global_ring.len() {
+            replication_factor = self.unique_nodes_in_global_ring.len();
+        }
+
+        self.global_ring
+            .ring_range(*token.borrow())
+            .unique()
+            .take(replication_factor)
+    }
+
+    /// Creates an iterator over network topology strategy replicas that live in a datacenter with
+    /// provided name.
+    pub fn nts_replicas_in_datacenter<'a>(
+        &'a self,
+        token: impl Borrow<Token>,
+        datacenter_name: &str,
+        mut replication_factor: usize,
+    ) -> impl Iterator<Item = &'a Arc<Node>> {
+        let dc_lb_data: &DatacenterNodes = match self.datacenters.get(datacenter_name) {
+            Some(dc_lb_data) => dc_lb_data,
+            None => &self.empty_dc_data,
+        };
+
+        if replication_factor > dc_lb_data.unique_nodes_in_dc_ring.len() {
+            replication_factor = dc_lb_data.unique_nodes_in_dc_ring.len();
+        }
+
+        let unique_dc_nodes = || dc_lb_data.dc_ring.ring_range(*token.borrow()).unique();
+        let double_dc_nodes = unique_dc_nodes().chain(unique_dc_nodes());
+
+        NtsReplicasInDatacenterIterator {
+            replicas_left_to_find: replication_factor,
+            unique_nodes_iter: double_dc_nodes,
+            used_racks: BTreeSet::new(),
+            acceptable_repeats: replication_factor.saturating_sub(dc_lb_data.rack_count),
+        }
+    }
+
+    pub fn unique_nodes_in_global_ring(&self) -> &[Arc<Node>] {
+        self.unique_nodes_in_global_ring.as_slice()
+    }
+
+    pub fn unique_nodes_in_datacenter_ring<'a>(
+        &'a self,
+        datacenter_name: &str,
+    ) -> Option<&'a [Arc<Node>]> {
+        self.datacenters
+            .get(datacenter_name)
+            .map(|dc| dc.unique_nodes_in_dc_ring.as_slice())
+    }
+}
+
+struct NtsReplicasInDatacenterIterator<'a, I>
+where
+    I: Iterator<Item = &'a Arc<Node>>,
+{
+    replicas_left_to_find: usize,
+    unique_nodes_iter: I,
+    used_racks: BTreeSet<Option<&'a str>>,
+    acceptable_repeats: usize,
+}
+
+impl<'a, I> Iterator for NtsReplicasInDatacenterIterator<'a, I>
+where
+    I: Iterator<Item = &'a Arc<Node>>,
+{
+    type Item = &'a Arc<Node>;
+
+    fn next(&mut self) -> Option<&'a Arc<Node>> {
+        if self.replicas_left_to_find == 0 {
+            return None;
+        }
+
+        let next_node: &Arc<Node> = match self.unique_nodes_iter.next() {
+            Some(next_node) => next_node,
+            None => return None,
+        };
+
+        let cur_rack: Option<&str> = next_node.rack.as_deref();
+        if self.used_racks.contains(&cur_rack) {
```

Comment:
Interesting thing here is that `BTreeSet::insert()` returns a bool instead of an Option, and it does not replace an old value with the new in case one was already present.